Sources/Commands/Workspace.swift

```diff
@@ -400,6 +412,20 @@ public class Workspace {
         try saveState()
     }
 
+    /// Creates a clone at the provided version and pins the repository at that version.
+    ///
+    /// - Parameters:
+    ///   - repository: The repository to pin. It should not be already pinned.
+    ///   - at: The version to pin at.
+    ///   - reason: The optional reason for pinning.
+    /// - Throws: WorkspaceOperationError, PinOperationError
+    func pin(repository: RepositorySpecifier, at version: Version, reason: String? = nil) throws {
+        // Clone the version we want to pin.
+        _ = try clone(specifier: repository, version: version)
```

Comment:
this is re-cloning the working checkout from the repositories cache, if the version is invalid this will throw (see L503) and will not pin or lose previous state. I figured re-cloning is better than checking out the pin version in the current checkout...this way is less messy (as opposed to resolving the version, getting the dependency, opening the handle, checking out the tag and updating the state which is what clone already does)